### ISSUE TITLE
Incorrect error report (E0308) on function taking two generic `IntoIterator`s 

### ISSUE THREAD

**Issue by cmrschwarz** [rust-lang/rust-analyzer#17257]
**rust-analyzer version**:  0.3.1958-standalone  (0.4.1960 affected aswell)

**rustc version**: rustc 1.78.0 (9b00956e5 2024-04-29)

**editor or extension**: VSCode v0.3.1958  ( v0.4.1960 pre-release affected aswell)

**relevant settings**:  N.A.

**repository link (if public, optional)**: N.A.

**code snippet to reproduce**: 
```rust
pub fn foo(
    bar: impl IntoIterator<Item = i32>,
    // removing this parameter miraculously 'fixes' the problem
    _: impl IntoIterator<Item = i32>,
) {
    let mut bar_iter = bar.into_iter();

    // shows as Option<{unknown}> instead of Option<i32>
    let type_confusion = bar_iter.next();

    // false error reported:
    // expected `Option<{unknown}>`, found `i32` (rust-analyzer E0308)
    let false_error = if let Some(v) = type_confusion {
        let v_cast = v as isize;
        v_cast.min(42)
    } else {
        42
    };
}
```

I tried my best to reduce this. 
Seems to me like two cascading bugs, first the unknown type, than the type error based on it.

Getting rid of this error for the stable VSCode release would be my plea here.
I would prefer to never receive **any** errors involving an *{unknown}* type.
Possibly as an option if other people have different tradeoffs?
I could imagine this only occuring because it is wrapped inside of the 'known' `Option` type (incorrectly),
but nontheless rust-analyzer knows that it messed up somewhere, so please leave me alone :laughing: .
 

Seing false positive type errors like these is quite bad for productivity during large refactorings, 
because it pollutes the list of remaining type errors in the IDE. 

A sincere thanks for all the work that you do. rust-analyzer is an amazing tool, and miles ahead of everything I've ever used in the C++ world. Big kudos to you guys.

**Comment by lnicola** [comment]
Duplicate of https://github.com/rust-lang/rust-analyzer/issues/10653.

**Comment by cmrschwarz** [comment]
I don't mind the breaking type inference, but why do I get an error? I'm sorry for wasting your time if this is indeed a duplicate.

**Comment by lnicola** [comment]
The error is because we think `min` is `Iterator::min`. We generally errors involving `{unknown}`, but not these ones, I guess, because they involve different types -- we're sure that `v_cast.min(42)` is an `Option<...>` that it's clear it can't be the same type as that `42` on the else branch.

Interestingly, though, `type_confusion` has the right type for me:

![image](https://github.com/rust-lang/rust-analyzer/assets/308347/b62a038a-8ca2-438a-844c-abc3198d16b6)

**Comment by cmrschwarz** [comment]
So I guess just an unfortunate consequence (actual hard error) of the same root cause as in #10653 ?
Fair enough then. Thanks for the clarification.

**Comment by lnicola** [comment]
You can silence the type mismatch on `false_error` using a type annotation (`let false_error: isize`), but we show the other one even with `isize::min(v as isize, 42isize)`, which seems unexpected. I wonder if there's another bug here..

**Comment by cmrschwarz** [comment]
> You can silence the type mismatch on `false_error` using a type annotation

Thank you so much for that! That makes a big difference for my workflow already.
